### ISSUE TITLE
Packer building a docker image within docker container

### ISSUE THREAD

**Issue by cprenzberg** [mitchellh/packer#4213]
Hi,

I am trying to use packer inside a docker container to build docker image as follows:

```
cat > bla.sh<<EOL
#!/bin/bash -xe


echo "Hello folks"

exit 0
EOL

chmod +x bla.sh

cat > manifest.json<<EOL
{
  "builders":[
      {
         "commit":true,
         "image":"ubuntu:14.04.5",
         "type":"docker"
      }
   ],
   "provisioners":[
      {
         "type":"shell",
         "scripts":[
            "bla.sh"
         ]
      }
   ],
   "post-processors":[
      [
         {
            "only":[
               "docker"
            ],
            "tag":"0.0.1",
            "type":"docker-tag",
            "repository":"docker-registry.example.com/ubuntu-base"
         },
         {
            "login_password":"foo",
            "login_email":"team@example.com",
            "only":[
               "docker"
            ],
            "login_server":"https://docker-registry.example.com",
            "login_username":"foo",
            "login":true,
            "type":"docker-push"
         }
      ],
      [{
            "only":[
               "docker"
            ],
            "tag":"latest",
            "type":"docker-tag",
            "repository":"docker-registry.example.com/ubuntu-base"
         },
         {
            "login_password":"foo",
            "login_email":"team@example.com",
            "only":[
               "docker"
            ],
            "login_server":"https://docker-registry.example.com",
            "login_username":"foo",
            "login":true,
            "type":"docker-push"
         }
      ]
   ]
}
EOL

packer validate manifest.json && PACKER_LOG=1 packer build -color=false manifest.json
```

However the provisioning gets stuck when trying to provision *bla.sh*. 

I run docker inside the container by mounting the host's /var/run/docker.sock into the container and just using the docker binaries available from the docker release repositories as follows:

```
docker run -it -v /var/run/docker.sock:/var/run/docker.sock --privileged --dns 10.10.12.6 -w /home/jenkins-agent/ docker-registry.example.com/jenkins-agent:0.0.7 bash
```

I can therefore push and pull and do all docker stuff within the container without problems. Here is the regular output from attempting the packer build specified in manifest.json from above

```
Template validated successfully.
docker output will be in this color.

==> docker: Creating a temporary directory for sharing data...
==> docker: Pulling Docker image: ubuntu:14.04.5
    docker: 14.04.5: Pulling from library/ubuntu
    docker: Digest: sha256:062bba17f92e749bd3092e7569aa06c6773ade7df603958026f2f5397431754c
    docker: Status: Image is up to date for ubuntu:14.04.5
==> docker: Starting docker container...
    docker: Run command: docker run -v /home/jenkins-agent/.packer.d/tmp/packer-docker731907216:/packer-files -d -i -t ubuntu:14.04.5 /bin/bash
    docker: Container ID: 24f8fa06db96506734222dcf792b7b2ee1816bb32e50fb0f5f5ded471b4cca36
==> docker: Provisioning with shell script: bla.sh
Build was aborted
```

When I look at the syslogs of the host, I see errors about **containerd** not being able to save the container. This error always appears whenever I try to run the packer build with a shell provisioner. The logs are as follows:

```
Nov 27 16:06:15 BER-LX-01 dockerd[13546]: time="2016-11-27T16:06:15.038205355+01:00" level=warning msg="containerd: unable to save 24f8fa06db96506734222dcf792b7b2ee1816bb32e50fb0f5f5ded471b4cca36:18e39664a701ed59ab176513c54ead3a98f9c65b97c40f7eafa42d95d4a1f529 starttime: open /proc/2127/stat: no such file or directory"
Nov 27 16:06:15 BER-LX-01 dockerd[13546]: time="2016-11-27T16:06:15.038266574+01:00" level=info msg="containerd: 24f8fa06db96506734222dcf792b7b2ee1816bb32e50fb0f5f5ded471b4cca36:18e39664a701ed59ab176513c54ead3a98f9c65b97c40f7eafa42d95d4a1f529 (pid 2127) has become an orphan, killing it"
Nov 27 16:06:17 BER-LX-01 dockerd[13546]: time="2016-11-27T16:06:17.110429824+01:00" level=warning msg="containerd: unable to save 24f8fa06db96506734222dcf792b7b2ee1816bb32e50fb0f5f5ded471b4cca36:266b063d3416af7c7c1450a3db71cd2708e5ad3907bd2e540159ca07f1a7c14f starttime: open /proc/2215/stat: no such file or directory"
Nov 27 16:06:17 BER-LX-01 dockerd[13546]: time="2016-11-27T16:06:17.110517715+01:00" level=info msg="containerd: 24f8fa06db96506734222dcf792b7b2ee1816bb32e50fb0f5f5ded471b4cca36:266b063d3416af7c7c1450a3db71cd2708e5ad3907bd2e540159ca07f1a7c14f (pid 2215) has become an orphan, killing it"
Nov 27 16:06:17 BER-LX-01 ntpd[1300]: Listen normally on 335 veth4b1b166 [fe80::80c6:22ff:fe17:1f9%625]:123
Nov 27 16:06:17 BER-LX-01 ntpd[1300]: new interface(s) found: waking up resolver
Nov 27 16:06:19 BER-LX-01 dockerd[13546]: time="2016-11-27T16:06:19.167605128+01:00" level=warning msg="containerd: unable to save 24f8fa06db96506734222dcf792b7b2ee1816bb32e50fb0f5f5ded471b4cca36:cf76c52ca2f0184d58f3292f958ce31de46e36c1efcf0f5100b2ed1b852d1f34 starttime: open /proc/2332/stat: no such file or directory"
Nov 27 16:06:19 BER-LX-01 dockerd[13546]: time="2016-11-27T16:06:19.167757064+01:00" level=info msg="containerd: 24f8fa06db96506734222dcf792b7b2ee1816bb32e50fb0f5f5ded471b4cca36:cf76c52ca2f0184d58f3292f958ce31de46e36c1efcf0f5100b2ed1b852d1f34 (pid 2332) has become an orphan, killing it"
Nov 27 16:06:21 BER-LX-01 dockerd[13546]: time="2016-11-27T16:06:21.208221241+01:00" level=warning msg="containerd: unable to save 24f8fa06db96506734222dcf792b7b2ee1816bb32e50fb0f5f5ded471b4cca36:db061d738df6e973753ebb376462fa178874bbbe8775bb17c895ae9ebc8d202c starttime: open /proc/2419/stat: no such file or directory"
Nov 27 16:06:21 BER-LX-01 dockerd[13546]: time="2016-11-27T16:06:21.208344182+01:00" level=info msg="containerd: 24f8fa06db96506734222dcf792b7b2ee1816bb32e50fb0f5f5ded471b4cca36:db061d738df6e973753ebb376462fa178874bbbe8775bb17c895ae9ebc8d202c (pid 2419) has become an orphan, killing it"
Nov 27 16:06:23 BER-LX-01 dockerd[13546]: time="2016-11-27T16:06:23.264529024+01:00" level=warning msg="containerd: unable to save 24f8fa06db96506734222dcf792b7b2ee1816bb32e50fb0f5f5ded471b4cca36:e19832ac3ae22645939999d5b27f313b51d81618faf2d25357263782674b5c18 starttime: open /proc/2507/stat: no such file or directory"
Nov 27 16:06:23 BER-LX-01 dockerd[13546]: time="2016-11-27T16:06:23.264670439+01:00" level=info msg="containerd: 24f8fa06db96506734222dcf792b7b2ee1816bb32e50fb0f5f5ded471b4cca36:e19832ac3ae22645939999d5b27f313b51d81618faf2d25357263782674b5c18 (pid 2507) has become an orphan, killing it"
Nov 27 16:06:25 BER-LX-01 dockerd[13546]: time="2016-11-27T16:06:25.304801578+01:00" level=warning msg="containerd: unable to save 24f8fa06db96506734222dcf792b7b2ee1816bb32e50fb0f5f5ded471b4cca36:4046c8aa6bc602535f4f94833dc1e640e965f18d094ca82e0510e3ed26bfee0a starttime: open /proc/2596/stat: no such file or directory"
Nov 27 16:06:25 BER-LX-01 dockerd[13546]: time="2016-11-27T16:06:25.305068372+01:00" level=info msg="containerd: 24f8fa06db96506734222dcf792b7b2ee1816bb32e50fb0f5f5ded471b4cca36:4046c8aa6bc602535f4f94833dc1e640e965f18d094ca82e0510e3ed26bfee0a (pid 2596) has become an orphan, killing it"
Nov 27 16:06:27 BER-LX-01 dockerd[13546]: time="2016-11-27T16:06:27.362358401+01:00" level=warning msg="containerd: unable to save 24f8fa06db96506734222dcf792b7b2ee1816bb32e50fb0f5f5ded471b4cca36:1a58ce35b24b7f59756bf9f1896fdc2d2b8074925efb017687ce7fb681025d24 starttime: open /proc/2683/stat: no such file or directory"
Nov 27 16:06:27 BER-LX-01 dockerd[13546]: time="2016-11-27T16:06:27.362476338+01:00" level=info msg="containerd: 24f8fa06db96506734222dcf792b7b2ee1816bb32e50fb0f5f5ded471b4cca36:1a58ce35b24b7f59756bf9f1896fdc2d2b8074925efb017687ce7fb681025d24 (pid 2683) has become an orphan, killing it"
Nov 27 16:06:29 BER-LX-01 dockerd[13546]: time="2016-11-27T16:06:29.418260661+01:00" level=warning msg="containerd: unable to save 24f8fa06db96506734222dcf792b7b2ee1816bb32e50fb0f5f5ded471b4cca36:d7af093901c4210660bde7973d882433616a5782bdc0743f4edcf18657a5dc6f starttime: open /proc/2776/stat: no such file or directory"
Nov 27 16:06:29 BER-LX-01 dockerd[13546]: time="2016-11-27T16:06:29.418387581+01:00" level=info msg="containerd: 24f8fa06db96506734222dcf792b7b2ee1816bb32e50fb0f5f5ded471b4cca36:d7af093901c4210660bde7973d882433616a5782bdc0743f4edcf18657a5dc6f (pid 2776) has become an orphan, killing it"
Nov 27 16:06:29 BER-LX-01 kernel: [685940.982746] docker0: port 8(veth4b1b166) entered forwarding state
Nov 27 16:06:31 BER-LX-01 dockerd[13546]: time="2016-11-27T16:06:31.494697239+01:00" level=warning msg="containerd: unable to save 24f8fa06db96506734222dcf792b7b2ee1816bb32e50fb0f5f5ded471b4cca36:0c7e6de2ee342c73a59c72dc945ca76fb597f4ed98f5b604ec92f5446e620331 starttime: open /proc/2864/stat: no such file or directory"
Nov 27 16:06:31 BER-LX-01 dockerd[13546]: time="2016-11-27T16:06:31.494792097+01:00" level=info msg="containerd: 24f8fa06db96506734222dcf792b7b2ee1816bb32e50fb0f5f5ded471b4cca36:0c7e6de2ee342c73a59c72dc945ca76fb597f4ed98f5b604ec92f5446e620331 (pid 2864) has become an orphan, killing it"
Nov 27 16:06:32 BER-LX-01 kernel: [685943.690146] docker0: port 8(veth4b1b166) entered disabled state
Nov 27 16:06:32 BER-LX-01 kernel: [685943.690183] vethd6868f9: renamed from eth0
Nov 27 16:06:32 BER-LX-01 kernel: [685943.762877] docker0: port 8(veth4b1b166) entered disabled state
Nov 27 16:06:32 BER-LX-01 kernel: [685943.764246] device veth4b1b166 left promiscuous mode
Nov 27 16:06:32 BER-LX-01 kernel: [685943.764249] docker0: port 8(veth4b1b166) entered disabled state
Nov 27 16:06:34 BER-LX-01 ntpd[1300]: Deleting interface #335 veth4b1b166, fe80::80c6:22ff:fe17:1f9%625#123, interface stats: received=0, sent=0, dropped=0, active_time=17 secs
```

**Host Platform**

```
root@41b8a538733d:/# docker version
Client:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   6b644ec
 Built:        Wed Oct 26 21:44:32 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   6b644ec
 Built:        Wed Oct 26 22:01:48 2016
 OS/Arch:      linux/amd64


root@41b8a538733d:/# packer version
Packer v0.12.0

```

Debug output log can be found at: https://gist.github.com/cprenzberg/27c53c4169c5c05ae8d88ff1d7ef657f

I look forward to any help you have to offer.

Best regards
Charles

**Comment by rickard-von-essen** [comment]
Guessing a bit, but it might be caused by the docker communicationor implementation. There is a PR to switch this to use the docker api. If you would like testing that would be appreciated. #3922

**Comment by cprenzberg** [comment]
Hi Rickard,

how may I test the code referenced in #3922 ?

Thanks

**Comment by cprenzberg** [comment]
I tried out the code from #3922 and it unfortunately does not solve my problem. Do you have any other guesses as to why I am having this issue using packer 0.12.0 from inside a docker 1.12.3-enabled container?

**Comment by cprenzberg** [comment]
symptoms look similar to those in #1767

**Comment by cprenzberg** [comment]
@mitchellh any ideas here? :)

**Comment by hvindin** [comment]
Just to add my experience with what I believe is a similar issue, I've had nothing but trouble trying to use shell (inline or script) provisioners with docker running packer 0.12.1 and docker 1.12.1. I ended up just installing ansible on the host and using that to provision. Sadly it does mean I need to start with a base docker container with sshd installed.

I'll try and reproduce my issue and see if it might be the same.

**Comment by rickard-von-essen** [comment]
I haven't had any time to look into this and I'm hesitating to spend time on it instead of working on #4186 .

**Comment by hvindin** [comment]
Hm, must have missed that, definitely a much nicer solution to make packer/docker integration much more robust. Hopefully it will resolve this issue

**Comment by mwhooker** [comment]
Going to close this until after #4186

Once that's in, if this is still a problem, feel free to open a new issue

**Comment by AgentO3** [comment]
I ran into this issue. I want to deploy jenkins in a docker container which needs the ability to build docker images with packer. I ended up working around this issue by mounting the host's docker socket in the jenkins container which allows me to connect to the host's docker daemon. I also mounted the /tmp/packer directory on the host to the jenkins container. In the builds in run in jenkins that require packer docker support I set the env var PACKER_TMP_DIR to /tmp/packer. When packer starts the sibling container it will be able to access the packer scripts from the /tmp/packer directory.
